include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,492 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <cassert>
+#include <coroutine>
+#include <exception>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <seastar/core/future.hh>
+
+// seastar::coroutine::experimental::async_generator is inspired by the C++23 proposal
+// P2502R2 (https://wg21.link/P2502R2), which introduced std::generator for
+// synchronous coroutine-based range generation.
+//
+// Similar to P2502R2's generator, seastar::coroutine::experimental::async_generator
+// * prioritizes storing references to yielded objects instead of copying them
+// * generates a range with iterators that yield values
+//
+// However, there are key differences in seastar::coroutine::experimental::generator:
+//
+// * Alocator support:
+//   Seastar's generator does not support the Allocator template parameter.
+//   Unlike Seastar uses its built-in allocator eliminating the need for
+//   additional flexibility.
+// * Asynchronous Operations:
+//   - generator::begin() is a coroutine, unlike P2502R2's synchronous approach
+//   - generator::iterator::operator++() is a coroutine
+//   - generator::iterator::operator++(int) is a coroutine
```

Comment:
Looking at flat_mutation_reader as a model, it can be used as

```c++
    while (auto value_opt = co_await reader()) {
        process(value);
    }
    /// end of stream
```

What would the equivalent look like here? Calling begin()/operator++() directly seem like a reduction in ease of use.

Reply:
Can we have: co_await (gen() | [] (auto val) {});
by overloading operator| ?

Reply:
or actually: while (co_await (gen() | func)) {}

Reply:
What would it do?

Reply:
Ah, it looks like a for_each thing.

We could do

```c++
    co_await coroutine::for_each(gen, lambda);
```

Though a while loop is more readable IMO.

Reply:
A while loop would require moving one value at a time from the generator buffer (if any) so to return it as an optional rvalue, right?
While for_each can call the lambda on buffered values that are in a circular buffer without having to move them.
I'm not sure if that could be significant to performance though, if moving the value is cheap.

Reply:
We could implement for_each if someone wants.

However, co_await is more general. It can be used with logic to switch paths.

```c++

    auto first = co_await gen();
    if (!first) {
         throw "empty stream";
    }
    if (first->val > 3) {
          while (auto rest = co_await gen()) {
               // one path
          } else {
               // another
          }
    }

```

Reply:
yes.

Also, can't for_each be a method of the generator additional to to call operator, like this?
```
co_await gen.for_each(lambda);
```